### ISSUE TITLE
AttributeError: 'MySQLDialect_pyodbc' object has no attribute '_connection_charset'

### ISSUE THREAD

**Issue by trankillity** [sqlalchemy/sqlalchemy#7966]
### Describe the bug
Attempting to connect to a MariaDB server using PyODBC so that we can use built in Windows connection credentials and it's throwing this error.

Tracing back the following error seems to show that the `self._detect_charset(connection)` function that defines `_connection_charset` is [not implemented](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/mysql/base.py#L3104) for MySQLDialect. This is called in the `initialize` function of the class.

### To Reproduce

```python
import sqlalchemy as sa

connx = sa.create_engine("mysql+pyodbc://?odbc_connect=DRIVER={MariaDB ODBC 3.1 Driver};SERVER=<server>;Trusted_Connection=yes;")

with connx.begin() as conn:
    sql_query = sa.text("select * from test.test")
    result = conn.execute(sql_query)
    data = result.fetchall()
    print(data)

connx.dispose()
```

### Error

```
Traceback (most recent call last):
  File "c:\mariatest\run.py", line 7, in <module>
    with connx.begin() as conn:
  File "C:\<packages>\sqlalchemy\engine\base.py", line 3062, in begin
    conn = self.connect(close_with_result=close_with_result)
  File "C:\<packages>\sqlalchemy\engine\base.py", line 3234, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File "C:\<packages>\sqlalchemy\engine\base.py", line 96, in __init__
    else engine.raw_connection()
  File "C:\<packages>\sqlalchemy\engine\base.py", line 3313, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File "C:\<packages>\sqlalchemy\engine\base.py", line 3280, in _wrap_pool_connect
    return fn()
  File "C:\<packages>\sqlalchemy\pool\base.py", line 310, in connect
    return _ConnectionFairy._checkout(self)
  File "C:\<packages>\sqlalchemy\pool\base.py", line 868, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "C:\<packages>\sqlalchemy\pool\base.py", line 476, in checkout
    rec = pool._do_get()
  File "C:\<packages>\sqlalchemy\pool\impl.py", line 145, in _do_get
    with util.safe_reraise():
  File "C:\<packages>\sqlalchemy\util\langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "C:\<packages>\sqlalchemy\util\compat.py", line 207, in raise_
    raise exception
  File "C:\<packages>\sqlalchemy\pool\impl.py", line 143, in _do_get
    return self._create_connection()
  File "C:\<packages>\sqlalchemy\pool\base.py", line 256, in _create_connection
    return _ConnectionRecord(self)
  File "C:\<packages>\sqlalchemy\pool\base.py", line 371, in __init__
    self.__connect()
  File "C:\<packages>\sqlalchemy\pool\base.py", line 679, in __connect
    )._exec_w_sync_on_first_run(self.dbapi_connection, self)
  File "C:\<packages>\sqlalchemy\event\attr.py", line 329, in _exec_w_sync_on_first_run
    self(*args, **kw)
  File "C:\<packages>\sqlalchemy\event\attr.py", line 343, in __call__
    fn(*args, **kw)
  File "C:\<packages>\sqlalchemy\util\langhelpers.py", line 1691, in go
    return once_fn(*arg, **kw)
  File "C:\<packages>\sqlalchemy\engine\create.py", line 686, in first_connect
    dialect.initialize(c)
  File "C:\<packages>\sqlalchemy\dialects\mysql\base.py", line 2710, in initialize
    self._connection_charset = self._detect_charset(connection)
  File "C:\<packages>\sqlalchemy\dialects\mysql\pyodbc.py", line 92, in _detect_charset
    value = self._fetch_setting(connection, "character_set_client")
  File "C:\<packages>\sqlalchemy\dialects\mysql\base.py", line 3114, in _fetch_setting
    charset = self._connection_charset
AttributeError: 'MySQLDialect_pyodbc' object has no attribute '_connection_charset'
``` 

### Versions

- OS: Windows 10
- Python: 3.10
- SQLAlchemy: 1.4.36
- Database: MariaDB
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): pyodbc

### Additional context

Looks like this may not have ever been implemented for MySQL, so I'm confused how it's worked in the past as the only definition of `self._connection_charset` seems to be by this function with no fallback?

**Comment by zzzeek** [comment]
so just like on #7871 ...I really would like someone to be available for fixing and testing.   we'd prefer not to maintain this pretty simple dialect.

**Comment by CaselIT** [comment]
Was this dialect removed on 2.0 like the other unsupported ones? Should it?

**Comment by zzzeek** [comment]
the poster has a use case which is they want to use windows authentication.   the dialect is very small.   i just am hesitant to installing mysql-pyodbc on our CI machines and getting every use case to work because pyodbc has all kinds of dumb datatype issues

**Comment by gordthompson** [comment]
> Attempting to connect to a MariaDB server using PyODBC so that we can use built in Windows connection credentials

I'm not a MariaDB expert, but I'm fairly confident that this is not possible using `Trusted_Connection=yes`. AFAIK that is a feature specific to Microsoft's ODBC driver for SQL Server (msodbcsql).

I just tried the following using plain pyodbc and the latest official release of MariaDB ODBC:

```python
import pyodbc

connection_string = (
    "Driver=Mariadb ODBC 3.1 Driver;"
    "Server=192.168.0.199;"
    "Port=3307;"
    "Database=test;"
    "Trusted_Connection=Yes;"
)
cnxn = pyodbc.connect(connection_string)

crsr = cnxn.cursor()
print(crsr.execute("SELECT 1 AS foo").fetchval())
```

That gave me

> pyodbc.InterfaceError: ('28000', "[28000] [ma-3.1.15]Access denied for user 'Gord'@'192.168.0.179' (using password: NO) (1045) (SQLDriverConnect); [28000] [ma-3.1.15]Access denied for user 'Gord'@'192.168.0.179' (using password: NO) (1045)")

At the MariaDB console I tried `CREATE USER 'Gord';` and that failed, but `CREATE USER 'gord'` succeeded. If I use

```python
connection_string = (
    "Driver=Mariadb ODBC 3.1 Driver;"
    "Server=192.168.0.199;"
    "Port=3307;"
    "Database=test;"
    "UID=gord;"
)
```

then that works, but it is not using Windows credentials (i.e., user 'gord' connects with no password).

Linux clients using msodbcsql (SQL Server) need to have Kerberos set up to connect with "Windows" credentials. I don't really know if MariaDB ODBC supports Kerberos, but my quick scan of the docs found no mention of it.

TL;DR - I suspect that using `mysql+pyodbc://` with MariaDB ODBC is not going to give you the functionality you desire, so you'd probably be better off using one of the supported DBAPI layers for your SQLAlchemy connections. Details [here](https://docs.sqlalchemy.org/en/14/dialects/mysql.html#mariadb-support).

**Comment by CaselIT** [comment]
From the documentation `Trusted_Connection` seems to be a thing in [db2](https://www.ibm.com/docs/en/db2/9.7?topic=keywords-trusted-connection) but there is no mention on [mysql](https://dev.mysql.com/doc/connector-odbc/en/connector-odbc-configuration-connection-parameters.html) or [mariadb](https://mariadb.com/kb/en/about-mariadb-connector-odbc/#parameters)

**Comment by trankillity** [comment]
Hi all, apologies for the delayed response on this. Yes, we have a use case for using integrated Windows auth to connect to the MariaDB. MariaDB can be configured using [authentication plugins](https://mariadb.com/kb/en/authentication-plugins/) so that it checks against a separate auth source for credentials. We've configured ours using GSSAPI for NTLM auth.

@gordthompson's suggested snipped **does** work using this auth plugin on my end.

I understand `pyodbc` may not be overly well supported, but I can't find any other connector that supports integrated Windows auth. PyMySQL apparently has a `auth_plugin_map` variable in the connection object, but can't find any docs on how it would be used.

**Comment by zzzeek** [comment]
OK well we have some time before the next release.

Please test this patch.   if you can fully connect and no more errors we can merge.

```diff
diff --git a/lib/sqlalchemy/dialects/mysql/pyodbc.py b/lib/sqlalchemy/dialects/mysql/pyodbc.py
index aa2190bf46..33e2d38648 100644
--- a/lib/sqlalchemy/dialects/mysql/pyodbc.py
+++ b/lib/sqlalchemy/dialects/mysql/pyodbc.py
@@ -88,6 +88,8 @@ class MySQLDialect_pyodbc(PyODBCConnector, MySQLDialect):
         #
         # If it's decided that issuing that sort of SQL leaves you SOL, then
         # this can prefer the driver value.
+
+        self._connection_charset = None
         try:
             value = self._fetch_setting(connection, "character_set_client")
             if value:

```

**Comment by trankillity** [comment]
That patch certainly gets me a step closer! Different error this time.
```
Traceback (most recent call last):
  File "c:\mariatest\run.py", line 7, in <module>
    with connx.begin() as conn:
  File "C:\<packages>\sqlalchemy\engine\base.py", line 3062, in begin
    conn = self.connect(close_with_result=close_with_result)
  File "C:\<packages>\sqlalchemy\engine\base.py", line 3234, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File "C:\<packages>\sqlalchemy\engine\base.py", line 96, in __init__
    else engine.raw_connection()
  File "C:\<packages>\sqlalchemy\engine\base.py", line 3313, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File "C:\<packages>\sqlalchemy\engine\base.py", line 3280, in _wrap_pool_connect
    return fn()
  File "C:\<packages>\sqlalchemy\pool\base.py", line 310, in connect
    return _ConnectionFairy._checkout(self)
  File "C:\<packages>\sqlalchemy\pool\base.py", line 868, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "C:\<packages>\sqlalchemy\pool\base.py", line 476, in checkout
    rec = pool._do_get()
  File "C:\<packages>\sqlalchemy\pool\impl.py", line 145, in _do_get
    with util.safe_reraise():
  File "C:\<packages>\sqlalchemy\util\langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "C:\<packages>\sqlalchemy\util\compat.py", line 207, in raise_
    raise exception
  File "C:\<packages>\sqlalchemy\pool\impl.py", line 143, in _do_get
    return self._create_connection()
  File "C:\<packages>\sqlalchemy\pool\base.py", line 256, in _create_connection
    return _ConnectionRecord(self)
  File "C:\<packages>\sqlalchemy\pool\base.py", line 371, in __init__
    self.__connect()
  File "C:\<packages>\sqlalchemy\pool\base.py", line 679, in __connect
    )._exec_w_sync_on_first_run(self.dbapi_connection, self)
  File "C:\<packages>\sqlalchemy\event\attr.py", line 329, in _exec_w_sync_on_first_run
    self(*args, **kw)
  File "C:\<packages>\sqlalchemy\event\attr.py", line 343, in __call__
    fn(*args, **kw)
  File "C:\<packages>\sqlalchemy\util\langhelpers.py", line 1691, in go
    return once_fn(*arg, **kw)
  File "C:\<packages>\sqlalchemy\engine\create.py", line 686, in first_connect
    dialect.initialize(c)
  File "C:\<packages>\sqlalchemy\dialects\mysql\base.py", line 2740, in initialize
    self._warn_for_known_db_issues()
  File "C:\<packages>\sqlalchemy\dialects\mysql\base.py", line 2745, in _warn_for_known_db_issues
    if mdb_version > (10, 2) and mdb_version < (10, 2, 9):
TypeError: '>' not supported between instances of 'NoneType' and 'tuple'
```
Looking into that error, it looks like the version isn't getting defined (`self._mariadb_normalized_version_info` is `None`). If I comment out [this conditional](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/mysql/base.py#L2720-L2729) then I can connect to the DB without incident.

**Comment by zzzeek** [comment]
let's skip the pyodbc server version routine

```diff
diff --git a/lib/sqlalchemy/dialects/mysql/pyodbc.py b/lib/sqlalchemy/dialects/mysql/pyodbc.py
index aa2190bf46..12011d70d1 100644
--- a/lib/sqlalchemy/dialects/mysql/pyodbc.py
+++ b/lib/sqlalchemy/dialects/mysql/pyodbc.py
@@ -88,6 +88,8 @@ class MySQLDialect_pyodbc(PyODBCConnector, MySQLDialect):
         #
         # If it's decided that issuing that sort of SQL leaves you SOL, then
         # this can prefer the driver value.
+
+        self._connection_charset = None
         try:
             value = self._fetch_setting(connection, "character_set_client")
             if value:
@@ -101,6 +103,9 @@ class MySQLDialect_pyodbc(PyODBCConnector, MySQLDialect):
         )
         return "latin1"
 
+    def _get_server_version_info(self, connection):
+        return MySQLDialect._get_server_version_info(self, connection)
+
     def _extract_error_code(self, exception):
         m = re.compile(r"\((\d+)\)").search(str(exception.args))
         c = m.group(1)

```

**Comment by trankillity** [comment]
Yep, that patch has allowed me to successfully connect and query the MariaDB database using `pyodbc` and `Trusted_Connection`.

I'm only new to programming and Python so I cannot contribute to maintaining the connector unfortunately, but more than happy to help test it.

**Comment by zzzeek** [comment]
Just needed you to test it, thanks.

when this merges to the "main" and "rel_1_4" branches, you will want to "git clone" our rel_1_4 branch and make sure everything works for you.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**bypass pyodbc default server version / set charset** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3821

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_4** branch:

**bypass pyodbc default server version / set charset** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3822
